src/scatter.h

```diff
@@ -33,222 +35,335 @@ template <class T = float> class FormFactorSphere {
+        // https://gcc.gnu.org/gcc-9/porting_to.html#ompdatasharing
+        // #pragma omp parallel for collapse(2) default(none) shared(directions, p_max, boxlength) shared(positions)
+        #pragma omp parallel for collapse(2) default(shared)
+        for (int i = 0; i < directions.size(); ++i) {
+            for (int p = 1; p <= p_max; ++p) {                                // loop over multiples of q
+                const Point q = (2 * pc::pi * p / boxlength) * directions[i]; // scattering vector
+                #ifdef __GNUC__
+                // When sine and cosine is computed in separate loops, advanced sine and cosine implementation
+                // utilizing SIMD instructions may be used to get at least 4 times performance boost.
+                // As of January 2020, only GCC exploits this using libmvec library if --ffast-math is enabled.
+                std::vector<T> qr_products(positions.size());
+                std::transform(positions.begin(), positions.end(), qr_products.begin(),
+                               [&q](auto &r) { return q.dot(r); });
+                // as of January 2020 the std::transform_reduce is not implemented in libc++
+                T sum_sin = 0;
+                for (auto &qr : qr_products) {
+                    sum_sin += std::sin(qr);
+                }
+                // as of January 2020 the std::transform_reduce is not implemented in libc++
+                T sum_cos = 0;
+                for (auto &qr : qr_products) {
+                    sum_cos += std::cos(qr);
+                }
+                #else
+                // TODO: Optimize also for other compilers than GCC by using a vector math library, e.g.,
+                // TODO: https://github.com/vectorclass/version2
+                T sum_sin = 0, sum_cos = 0;
+                for (auto &r : positions) { // loop over positions
+                    T qr = _q.dot(r);       // scalar product q*r
                     sum_sin += sin(qr);
                     sum_cos += cos(qr);
                 }
+                #endif
                 // collect average, `norm()` gives the scattering vector length
-                S[_q.norm()] += (sum_sin * sum_sin + sum_cos * sum_cos) / double(positions.size());
-                W[_q.norm()] += 1.0;
+                const T sf = (sum_sin * sum_sin + sum_cos * sum_cos) / (float)(positions.size());
+                #pragma omp critical
+                // avoid race conditions when updating the map
+                addSampling(q.norm(), sf, 1.0);
             }
-        };
-        std::for_each(directions.begin(), directions.end(), func); // add std::execution::par policy?
+        }
     }
 
-    void save(const std::string &filename) {
-        if (not S.empty()) {
-            std::ofstream f(filename.c_str());
-            if (f)
-                for (auto &i : S)
-                    f << i.first << " " << i.second / W[i.first] << "\n";
+    int getQMultiplier() {
```

Comment:
is this for a future purpose?

Reply:
It is used in `analysis.cpp` in `to_json` function.